src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepExecution.java

```diff
@@ -23,83 +23,108 @@
 import org.jenkinsci.plugins.workflow.graphanalysis.LinearScanner;
-import org.jenkinsci.plugins.workflow.steps.AbstractSynchronousStepExecution;
-import org.jenkinsci.plugins.workflow.steps.FlowInterruptedException;
 import org.jenkinsci.plugins.workflow.steps.StepContext;
-import org.jenkinsci.plugins.workflow.steps.StepContextParameter;
-
-import com.google.inject.Inject;
+import org.jenkinsci.plugins.workflow.steps.SynchronousStepExecution;
 
-import hudson.AbortException;
-import hudson.Extension;
-import hudson.model.Executor;
-import hudson.model.InvisibleAction;
-import hudson.model.Job;
-import hudson.model.Result;
-import hudson.model.Run;
-import hudson.model.TaskListener;
-import hudson.model.listeners.RunListener;
-import jenkins.model.Jenkins;
-
-public class MilestoneStepExecution extends AbstractSynchronousStepExecution<Void> {
+public class MilestoneStepExecution extends SynchronousStepExecution<Void> {
 
+    private static final Predicate<FlowNode> ORDINAL_MATCHER = FlowScanningUtils.hasActionPredicate(OrdinalAction.class);
     private static final Logger LOGGER = Logger.getLogger(MilestoneStepExecution.class.getName());
+    private final String label;
+    private final Integer ordinal;
+    private final boolean unsafe;
+
+    public MilestoneStepExecution(@NonNull StepContext context, @CheckForNull String label, @CheckForNull Integer ordinal, boolean unsafe) {
+        super(context);
+        this.label = label;
+        this.ordinal = ordinal;
+        this.unsafe = unsafe;
+    }
 
-    @Inject(optional=true) private transient MilestoneStep step;
-    @StepContextParameter private transient Run<?,?> run;
-    @StepContextParameter private transient FlowNode node;
-    @StepContextParameter private transient TaskListener listener;
+    /**
+     * @param buildNumber The build number currently querying for builds to cancel
+     * @param ordinal The ordinal the build just passed. {@code null} means it just started.
+     * @param milestones A map keyed by build numbers recording their current milestone.
+     * @return A subset of build numbers among the given milestones eligible for cancellation.
+     */
+    public static Map<Integer, Integer> getBuildsToCancel(int buildNumber, @CheckForNull Integer ordinal, @NonNull Map<Integer, Integer> milestones) {
+        Map<Integer, Integer> result = new HashMap<>();
+        for (Map.Entry<Integer, Integer> entry : milestones.entrySet()) {
+            if (entry.getKey() < buildNumber) {
+                if (entry.getValue() == null || (ordinal != null && entry.getValue() < ordinal)) {
+                    Integer key = entry.getKey();
+                    result.put(key, buildNumber);
+                }
+            } else if (entry.getKey() > buildNumber && ((ordinal == null && entry.getValue() != null) || (ordinal != null  && entry.getValue() != null && entry.getValue() >= ordinal))) {
+                // Defensive, this should never happen.
```

Comment:
I could imagine that in the case of an async resume, given `m < n` If build `#m` gets resumed later than `#n`, and `#n` has progressed to the same milestone as `#m`, then `#m` would hit his branch on its resume.